ignition test sometimes fails

### Steps to reproduce

Sometimes the ignition overlay test fails. Then re-running the test succeeds.

There appears to be some kind of timing / race condition in the test, or maybe in the overlay.

I'm capturing some output I've discovered from previous failures here.

### Error message

This is the difference I saw between expected and actual:

```
map[ignition:map[version:3.1.0] storage:map[disks:[map[device:/dev/vda partitions:[map[label:scratch shouldExist:true wipePartitionEntry:false] map[label:swap number:1 shouldExist:false sizeMiB:1024 wipePartitionEntry:false]] wipeTable:true]] filesystems:[map[device:/dev/disk/by-partlabel/scratch format:btrfs path:/scratch wipeFilesystem:true] map[device:/dev/disk/by-partlabel/swap format:swap path:swap wipeFilesystem:false]]]]

map[ignition:map[version:3.1.0] storage:map[disks:[map[device:/dev/vda partitions:[map[label:scratch shouldExist:true wipePartitionEntry:false] map[label:swap number:1 shouldExist:false sizeMiB:1024 wipePartitionEntry:false]] wipeTable:true]] filesystems:[map[device:/dev/disk/by-partlabel/swap format:swap path:swap wipeFilesystem:false] map[device:/dev/disk/by-partlabel/scratch format:btrfs path:/scratch wipeFilesystem:true]]]]
```

### Information on your system

Some recent `main` build.

```
[janderson@rocky ~]$ cat /etc/os-release 
NAME="Rocky Linux"
VERSION="9.4 (Blue Onyx)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="9.4"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Rocky Linux 9.4 (Blue Onyx)"
ANSI_COLOR="0;32"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:rocky:rocky:9::baseos"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
SUPPORT_END="2032-05-31"
ROCKY_SUPPORT_PRODUCT="Rocky-Linux-9"
ROCKY_SUPPORT_PRODUCT_VERSION="9.4"
REDHAT_SUPPORT_PRODUCT="Rocky Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="9.4"
```

### General information

- [X] I have run `wwctl version` and reported the contents of `/etc/os-release`
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have captured and reported relevant error messages and logs

Capturing another instance, just in case they're different.

```
--- FAIL: Test_ignitionOverlay (0.01s)
    --- FAIL: Test_ignitionOverlay/ignition:ignition.json (0.00s)
        ignition_test.go:67: 
            	Error Trace:	/home/janderson/warewulf.git/tree/main/overlays/ignition/internal/ignition_test.go:67
            	Error:      	Should be true
            	Test:       	Test_ignitionOverlay/ignition:ignition.json
            	Messages:   	expected: map[ignition:map[version:3.1.0] storage:map[disks:[map[device:/dev/vda partitions:[map[label:scratch shouldExist:true wipePartitionEntry:false] map[label:swap number:1 shouldExist:false sizeMiB:1024 wipePartitionEntry:false]] wipeTable:true]] filesystems:[map[device:/dev/disk/by-partlabel/scratch format:btrfs path:/scratch wipeFilesystem:true] map[device:/dev/disk/by-partlabel/swap format:swap path:swap wipeFilesystem:false]]]]
            	            	actual  : map[ignition:map[version:3.1.0] storage:map[disks:[map[device:/dev/vda partitions:[map[label:scratch shouldExist:true wipePartitionEntry:false] map[label:swap number:1 shouldExist:false sizeMiB:1024 wipePartitionEntry:false]] wipeTable:true]] filesystems:[map[device:/dev/disk/by-partlabel/swap format:swap path:swap wipeFilesystem:false] map[device:/dev/disk/by-partlabel/scratch format:btrfs path:/scratch wipeFilesystem:true]]]]
```

I found `assert.JSONEq` and  now I'm getting an even clearer error message.

```
[janderson@rocky main]$ go test overlays/ignition/internal/ignition_test.go
--- FAIL: Test_ignitionOverlay (0.01s)
    --- FAIL: Test_ignitionOverlay/ignition:ignition.json (0.00s)
        ignition_test.go:62: 
            	Error Trace:	/home/janderson/warewulf.git/tree/main/overlays/ignition/internal/ignition_test.go:62
            	Error:      	Not equal: 
            	            	expected: map[string]interface {}{"ignition":map[string]interface {}{"version":"3.1.0"}, "storage":map[string]interface {}{"disks":[]interface {}{map[string]interface {}{"device":"/dev/vda", "partitions":[]interface {}{map[string]interface {}{"label":"scratch", "shouldExist":true, "wipePartitionEntry":false}, map[string]interface {}{"label":"swap", "number":1, "shouldExist":false, "sizeMiB":1024, "wipePartitionEntry":false}}, "wipeTable":true}}, "filesystems":[]interface {}{map[string]interface {}{"device":"/dev/disk/by-partlabel/scratch", "format":"btrfs", "path":"/scratch", "wipeFilesystem":true}, map[string]interface {}{"device":"/dev/disk/by-partlabel/swap", "format":"swap", "path":"swap", "wipeFilesystem":false}}}}
            	            	actual  : map[string]interface {}{"ignition":map[string]interface {}{"version":"3.1.0"}, "storage":map[string]interface {}{"disks":[]interface {}{map[string]interface {}{"device":"/dev/vda", "partitions":[]interface {}{map[string]interface {}{"label":"scratch", "shouldExist":true, "wipePartitionEntry":false}, map[string]interface {}{"label":"swap", "number":1, "shouldExist":false, "sizeMiB":1024, "wipePartitionEntry":false}}, "wipeTable":true}}, "filesystems":[]interface {}{map[string]interface {}{"device":"/dev/disk/by-partlabel/swap", "format":"swap", "path":"swap", "wipeFilesystem":false}, map[string]interface {}{"device":"/dev/disk/by-partlabel/scratch", "format":"btrfs", "path":"/scratch", "wipeFilesystem":true}}}}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -27,2 +27,8 @@
            	            	    (map[string]interface {}) (len=4) {
            	            	+    (string) (len=6) "device": (string) (len=27) "/dev/disk/by-partlabel/swap",
            	            	+    (string) (len=6) "format": (string) (len=4) "swap",
            	            	+    (string) (len=4) "path": (string) (len=4) "swap",
            	            	+    (string) (len=14) "wipeFilesystem": (bool) false
            	            	+   },
            	            	+   (map[string]interface {}) (len=4) {
            	            	     (string) (len=6) "device": (string) (len=30) "/dev/disk/by-partlabel/scratch",
            	            	@@ -31,8 +37,2 @@
            	            	     (string) (len=14) "wipeFilesystem": (bool) true
            	            	-   },
            	            	-   (map[string]interface {}) (len=4) {
            	            	-    (string) (len=6) "device": (string) (len=27) "/dev/disk/by-partlabel/swap",
            	            	-    (string) (len=6) "format": (string) (len=4) "swap",
            	            	-    (string) (len=4) "path": (string) (len=4) "swap",
            	            	-    (string) (len=14) "wipeFilesystem": (bool) false
            	            	    }
            	Test:       	Test_ignitionOverlay/ignition:ignition.json
FAIL
FAIL	command-line-arguments	0.022s
FAIL
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -32,6 +32,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Bump github.com/coreos/ignition/v2 from 2.15.0 to 2.19.0 #1239
 - Disable building containers by default when calling `wwctl container copy`. #1378
 - Split wwinit and generic overlays into discrete functionality. #987
+- Updated IgnitionJson to sort filesystems. #1433
 
 ### Fixed
 

--- a/internal/pkg/node/ignition.go
+++ b/internal/pkg/node/ignition.go
@@ -47,6 +47,9 @@ func (node *NodeInfo) GetStorage() (stor types_3_4.Storage, err error, rep strin
 		wwlog.Debug("created file system struct: %v", myFs)
 		fileSystems = append(fileSystems, myFs)
 	}
+	sort.SliceStable(fileSystems, func(i int, j int) bool {
+		return fileSystems[i].Device < fileSystems[j].Device
+	})
 	var disks []types_3_4.Disk
 	for diskDev, disk := range node.Disks {
 		var partitions []types_3_4.Partition

--- a/overlays/ignition/internal/ignition_test.go
+++ b/overlays/ignition/internal/ignition_test.go
@@ -2,8 +2,6 @@ package ignition
 
 import (
 	"bytes"
-	"encoding/json"
-	"reflect"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -61,10 +59,7 @@ func Test_ignitionOverlay(t *testing.T) {
 			assert.Empty(t, stdout.String())
 			assert.Empty(t, stderr.String())
 			if tt.json {
-				var expected, actual interface{}
-				assert.NoError(t, json.Unmarshal([]byte(tt.log), &expected))
-				assert.NoError(t, json.Unmarshal(logbuf.Bytes(), &actual))
-				assert.True(t, reflect.DeepEqual(expected, actual), "expected: %v\nactual  : %v", expected, actual)
+				assert.JSONEq(t, tt.log, logbuf.String())
 			} else {
 				assert.Equal(t, tt.log, logbuf.String())
 			}
@@ -116,5 +111,44 @@ What=/dev/disk/by-partlabel/swap
 RequiredBy=swap.target
 `
 
-const ignition_json string = `{"ignition":{"version":"3.1.0"},"storage":{"disks":[{"device":"/dev/vda","partitions":[{"label":"scratch","shouldExist":true,"wipePartitionEntry":false},{"label":"swap","number":1,"shouldExist":false,"sizeMiB":1024,"wipePartitionEntry":false}],"wipeTable":true}],"filesystems":[{"device":"/dev/disk/by-partlabel/scratch","format":"btrfs","path":"/scratch","wipeFilesystem":true},{"device":"/dev/disk/by-partlabel/swap","format":"swap","path":"swap","wipeFilesystem":false}]}}
-`
+const ignition_json string = `{
+  "ignition": {
+    "version": "3.1.0"
+  },
+  "storage": {
+    "disks": [
+      {
+        "device": "/dev/vda",
+        "partitions": [
+          {
+            "label": "scratch",
+            "shouldExist": true,
+            "wipePartitionEntry": false
+          },
+          {
+            "label": "swap",
+            "number": 1,
+            "shouldExist": false,
+            "sizeMiB": 1024,
+            "wipePartitionEntry": false
+          }
+        ],
+        "wipeTable": true
+      }
+    ],
+    "filesystems": [
+      {
+        "device": "/dev/disk/by-partlabel/scratch",
+        "format": "btrfs",
+        "path": "/scratch",
+        "wipeFilesystem": true
+      },
+      {
+        "device": "/dev/disk/by-partlabel/swap",
+        "format": "swap",
+        "path": "swap",
+        "wipeFilesystem": false
+      }
+    ]
+  }
+}`

